netket/hilbert/tensor_hilbert_discrete.py

```diff
@@ -58,18 +58,29 @@ def __init__(self, *hilb_spaces: DiscreteHilbert):
             )
 
         shape = np.concatenate([hi.shape for hi in hilb_spaces])
-
+        self._initialized = False
         super().__init__(hilb_spaces, shape=shape)
 
-        # pre-compute indexing data iff the tensor space is still indexable
-        if all(hi.is_indexable for hi in hilb_spaces) and _is_indexable(shape):
-            self._ns_states = [hi.n_states for hi in self._hilbert_spaces]
-            self._ns_states_r = np.flip(self._ns_states)
-            self._cum_ns_states = np.concatenate([[0], np.cumprod(self._ns_states)])
-            self._cum_ns_states_r = np.flip(
-                np.cumprod(np.concatenate([[1], np.flip(self._ns_states)]))[:-1]
-            )
-            self._n_states = np.prod(self._ns_states)
+    @property
+    def is_indexable(self) -> bool:
+        """Whether the space can be indexed with an integer"""
+        return all(hi.is_indexable for hi in self._hilbert_spaces) and _is_indexable(
+            self.shape
+        )
+
+    def _computue_indexing_data(self):
```

Comment:
Idea: should we use `_setup` for consistency with the operators? If you prefer not, do correct the typo: to `_compute_indexing_data`

```suggestion
    def _setup(self):
```